### PR TITLE
feature(wrh): Add Harmony Dream loss balance in MuZero

### PR DESCRIPTION
**Pull request by ruiheng123**
Harmony Dream: Dynamically balance the weights of different loss functions during training. 
For more details, refer to original paper: https://arxiv.org/abs/2310.00344

### CODE REVIEW

File: lzero/model/muzero_model.py [hunk 1]
```diff
@@ -182,7 +184,16 @@ def __init__(
             activation=activation,
             norm_type=norm_type
         )
-
+        
+        # Referred as https://github.com/thuml/HarmonyDream/blob/main/wmlib-torch/wmlib/agents/dreamerv2.py#L116
+        self.use_harmony = use_harmony
+        if self.use_harmony:
+            self.harmony_s1 = nn.Parameter(-torch.log(torch.tensor(1.0)))
```
Comment (puyuan1996) [inline lzero/model/muzero_model.py:191]: 这里初始化为为-torch.log(torch.tensor(1.0))，有什么特别的原因呢？为什么要加上负号呢？可以提个issue询问一下

File: lzero/policy/muzero.py [hunk 2]
```diff
@@ -297,7 +297,21 @@ def _init_learn(self) -> None:
         self.inverse_scalar_transform_handle = InverseScalarTransform(
             self._cfg.model.support_scale, self._cfg.device, self._cfg.model.categorical_distribution
         )
-
+        
+        # ==============================================================
```
Comment (puyuan1996) [inline lzero/policy/muzero.py:301]: 或许可以将 `use_harmony` 换成 `harmony_balance`, 并在muzero上方的默认config中加上harmony_balance 的默认为False，以及相关的注释

File: lzero/model/muzero_model.py [hunk 3]
```diff
@@ -182,7 +184,16 @@ def __init__(
...
+            self.harmony_s2 = nn.Parameter(-torch.log(torch.tensor(1.0)))
+            self.harmony_s3 = nn.Parameter(-torch.log(torch.tensor(1.0)))
+            self.harmony_s4 = nn.Parameter(-torch.log(torch.tensor(1.0)))
+            self.harmony_s5 = nn.Parameter(-torch.log(torch.tensor(1.0)))
```
Comment (puyuan1996) [inline lzero/model/muzero_model.py:195]: s1-5换成对应loss的name简写，例如 harmony_latent, harmony_policy, harmony_value, harmony_reward, harmony_entropy

Comment (puyuan1996) [reply lzero/model/muzero_model.py:195]: 可以类似下面这样写，方便扩展哈
```python
            # List of parameter names
            harmony_names = ["harmony_latent", "harmony_policy", "harmony_value", "harmony_reward", "harmony_entropy"]
            
            # Initialize and name each parameter
            for name in harmony_names:
                param = nn.Parameter(-torch.log(torch.tensor(1.0)))
                setattr(self, name, param)
```

Comment (puyuan1996) [reply lzero/model/muzero_model.py:195]: 另外，这里的相关定义，应该移动到 muzero policy 的 _init_learn 中去哈，这样就可以避免下面这样的冗余代码，`self.harmony_s1 = self._model.harmony_s1`

**Review by puyuan1996 — CHANGES_REQUESTED** [review]

File: lzero/model/muzero_model.py (hunk 1 above)
Comment (ruiheng123) [reply lzero/model/muzero_model.py:191]: 图片
参照的原官方代码

File: zoo/atari/config/atari_muzero_config.py [hunk 4]
```diff
@@ -60,6 +60,7 @@
             use_sim_norm=True,
             use_sim_norm_kl_loss=False,
             model_type='conv',
+            harmony_balance=False # Referred as https://github.com/thuml/HarmonyDream.
```
Comment (puyuan1996) [inline zoo/atari/config/atari_muzero_config.py:63]: muzero默认是Flase，这里可以删除了哈

File: lzero/policy/muzero.py [hunk 5]
```diff
@@ -523,12 +536,33 @@ def _forward_learn(self, data: Tuple[torch.Tensor]) -> Dict[str, Union[float, in
...
+                + (reward_loss.mean() / torch.exp(self.harmony_reward))
+                # + (policy_entropy_loss.mean() / torch.exp(self.harmony_entropy))
+            ) 
+            weighted_total_loss = loss.mean()
+            weighted_total_loss += (
+                # torch.log(torch.exp(self.harmony_dynamics) + 1) +
+                torch.log(torch.exp(self.harmony_policy) + 1) +
+                torch.log(torch.exp(self.harmony_value) + 1) + 
+                torch.log(torch.exp(self.harmony_reward) + 1) 
+                # torch.log(torch.exp(self.harmony_entropy) + 1)
```
Comment (puyuan1996) [inline lzero/policy/muzero.py:557]: 把549，553，557行的harmony_dynamics和harmony_entropy相关的注释删除一下

File: lzero/policy/muzero.py [hunk 6]
```diff
@@ -70,6 +70,8 @@ class MuZeroPolicy(Policy):
             analysis_sim_norm=False,
             # (bool) Whether to analyze dormant ratio.
             analysis_dormant_ratio=False,
+            # (bool) Whether to use HarmonyDream to balance weights between different losses. Default to False.
+            harmony_balance=False
```
Comment (puyuan1996) [inline lzero/policy/muzero.py:74]: 增加HarmonyDream论文的参考，例如 More details can be found in <>.

File: lzero/policy/muzero.py [hunk 7]
```diff
@@ -523,12 +536,33 @@ def _forward_learn(self, data: Tuple[torch.Tensor]) -> Dict[str, Union[float, in
...
         # weighted loss with masks (some invalid states which are out of trajectory.)
-        loss = (
-                self._cfg.ssl_loss_weight * consistency_loss + self._cfg.policy_loss_weight * policy_loss +
-                self._cfg.value_loss_weight * value_loss + self._cfg.reward_loss_weight * reward_loss +
-                self._cfg.policy_entropy_loss_weight * policy_entropy_loss
-        )
-        weighted_total_loss = (weights * loss).mean()
+        # Nan appear when consistency loss or policy entropy loss uses harmony parameter as coefficient.
+        
+        # Referred as https://github.com/thuml/HarmonyDream/blob/main/wmlib-torch/wmlib/agents/dreamerv2.py#L161
```
Comment (puyuan1996) [inline lzero/policy/muzero.py:541]: 这行修改为：Please refer to *

**Review by puyuan1996 — CHANGES_REQUESTED** [review]
